[Release] Version 1.0.0 (#1)
luizmendez/learning-react-with-dogs: README.md

As Redux' actions are not asynchronous out of the box, the [redux-thunk](https://github.com/reduxjs/redux-thunk) middleware was used for resolving this behavior.

In this app, Redux handles the fetching and sending of data from the APIs as well as managing this information in the store. It also stores and removes messages from the message list that `<MessageHandler />` uses.

### Key parts

#### `<App />`

The root component of the App.
Connects with the Redux store and actions to be passed as props.
On mount, the component calls a Redux action that retrieves the dog's list from local storage if available, or fetches it from the API to be set on store.

#### Router

The `<Router />` component was coded from scratch based on the [react-router](https://github.com/ReactTraining/react-router) library. It has four main components: `<Router />`, `<Route />`, `<Redirect />` and `<Otherwise/ >`. There's also a `<Link />` component to assist with navigation.
Uses **React Context** to handle the state and pass it to components that may need it.

[Read more about the router](./components.md#router)

#### ShowDogs Components

> The main content of the app. Several components that fetch a list of dog breeds and manages image display of them.

Reviewer: ```suggestion
The main content of the app. Several components that fetch a list of dog breeds and manages displaying their images.
```

Author: 🍤

> Shows as a card list, with the image of each dog. On hover, a tooltip created with **React Portals** that follows the cursor with the breed name until it leaves the image.

Reviewer: ```suggestion
Shows as a card list, with the image of each dog. On hover, a tooltip created with **React Portals**  follows the cursor with the breed name until it leaves the image.
```

Author: 🍤

#### `<SubmitDog />`

**At the moment this component is still under work and only showcases error handling**

Allows the user to submit a dog image with the name and breed of the dog. The `<select>` element lists all the dogs' breeds received from the API and passed down by `<App />`.
An image preview is shown next to the form after the user selects an image from their computer.
Form submission calls an action that sends the form data to the corresponding endpoint, which currently don't exists and will be developed in the second part of this app. For now, it just showcases error handling.

[Read more about SubmitDog](./components.md#submitDog)

#### Error Handling

To display alerts, warning, errors and any necessary type of message, the app uses `<MessageHandler />` and `<Snackbar />` components.

The `<MessageHandler>` receives a list of messages which renders as `<Snackbars />` until they are either manually or automatically closed. If there are several messages at the same time they start to stack up from bottom to top. When the `<Snackbar />` component is closed, a callback function can be called to perform any other action, for example, removing the message from state.

> The `<Snackbar />` component shows either a string message or any given children. The component alone is able to autoclose after some time, defaults to 3000ms but can also be passed as a prop for different times. There are other options like not autoclosing, hiding the close button, and the style to be use for the message box (alert, warning, error).

Reviewer: ```suggestion
The `<Snackbar />` component shows either a string message or any given children. The component alone is able to autoclose after some time, defaults to 3000ms but can also be passed as a prop for different times. There are other options like not autoclosing, hiding the close button, and setting the style to be used for the message box (alert, warning, error).
```

Author: 🍤

[Read more about the Error Handling components](./components.md#error-handling)

### Disclaimer

Some of the app features may be incomplete or can be enhanced. The focus of this project was not to have a polished off application, but to learn, comprehend and implement modern tools and processes.

## Testing

Testing was implemented on a late stage of the UI development, in parallel to Redux. All of Redux' actions and reducers were implemented using **Test Driven Development**. The `<SubmitDog />` component was tested after learning how to unit test on Redux. After this, the `<MessageHandler />` and `<Snackbar />` components were created using TDD.

As testing was implemented late in the project, not all components have tests.

> To test the app in the project folder run `yarn test`.

Reviewer: ```suggestion
To test the app, in the project folder run `yarn test`.
```
But: To test the app, run `yarn test` in the root directory would be better

Author: 🍤

-   **Mock libraries** used:
    -   [mock-local-storage](https://www.npmjs.com/package/jest-localstorage-mock) mock for testing localstorage
    -   [mock-fetch](https://www.npmjs.com/package/fetch-mock) mock for testing dog list and dog image fetching
    -   [mock-store](https://github.com/dmitry-zaets/redux-mock-store) redux store mock for async actions and middleware

## Components

[See a detailed list of the components used in the project.](./components.md)

## Styles

-   Styles were added using a vanilla stylesheet that loads inside `<App />`.
-   [Bootstrap 4](https://getbootstrap.com/) is loaded from official CDN on index.html

## Conclusion

The project taught me how to use modern tools and processes such as **React**, **Redux**, **Test Driven Development** using **Jest** as a testing platform and in the next step of the project I will learn **NodeJS**.

The project was started without a boilerplate, I learned how to start up a project from scratch using **Yarn**, **Webpack** and **Babel**. This project was focused on React but this knowledge can be used for any kind of project.

> Git is used, which I have already knew and used before, but this is the first time I use the **git-flow** workflow in a project.

Reviewer: ```suggestion
Git is used, which I already knew and used before, but this is the first time I use the **git-flow** workflow in a project.
```

Author: 🍤

After configuring the project, I learned how to plan the architecture of an app, defining specs and features and how they should interact with the application and between them.

As the coding started, there was so much to learn in react, starting with JSX, components and their lifecycle, props, state, the render method, render props, high order components, refs, portals, context... to name some. My mindset also changed as I have to start thinking in React. Immutability, components just serving one purpose, when and when not to create a state, how components should communicate between each other.

> When the app state management necessities grew, Redux was implemented with Test Driven Development process using Jest as the testing platform. This was new for me, I learned a lot from this: as to how to set up a testing environment, which helpers I would need to use and how to use them, and what to test in the code. Then some components were created using TDD with the help of the **Enzyme** library.

Reviewer: ```suggestion
When the app state management necessities grew, Redux was implemented with Test Driven Development process using Jest as the testing platform. This was new for me, I learned a lot from this: like how to set up a testing environment, which helpers I would need to use and how to use them, and what to test in the code. Then some components were created using TDD with the help of the **Enzyme** library.
```

Author: 🍤

> Redux actions were used to fetch information from an API, so I learned to use **Thunk** as a middleware in these cases. I learned what actions and reducers are, and how they connect to the app, passing the store and actions as props to the components that needed them.

Reviewer: I'd like to think you learned how to use middleware, using Thunk as an example.

Author: 🍤

# Learning React with Dogs

### What is this?

As the name implies, this is a personal project, meant for learning while coding.
The main focus was to learn React, Redux, Test Driven Development with Jest and as a final step NodeJs.

> **At this part of the project, NodeJs is still not implemented but will be as next step.**

Reviewer: ```suggestion
**At this moment, NodeJS is still not implemented, but will be done soon.**
```

Author: 🍌

> ### What is the app's purpose?

Reviewer: I don't think this section is needed, it doesn't even really answer the question, it just lists some features. The question, what is the app's purpose? is answered in the What is this? section. I think you could make a high level overview of what the app does in that section, as well as why you built all this.

**TL;DR:** remove this section, most features are listed in depth in **The Project** section. Merge this section with What is this?

Author: done

> The app fetches data from [DogAPI](https://dog.ceo/dog-api/) and shows it in a simple way as an images list.

Reviewer: ```suggestion
The app fetches data from [DogAPI](https://dog.ceo/dog-api/) and shows it in a simple way as a list of images.
```

Author: 🍤

> The list can be filtered, contains a lazy load component for the dogs images and you can submit a dog picture through a form.

Reviewer: ```suggestion
The list can be filtered, it contains a lazy load component for the dog images and users can submit a dog picture through a form.
```

Author: 🍤

> It implements a handmade router and manages state with Redux and Context.

Reviewer: ```suggestion
The application's state is managed with Redux [INSERT HERE WHY]
The project uses its own router implementation based on [react-router](https://github.com/ReactTraining/react-router), which manages its state with React Context.
```

Author: 🍤

> For a more detailed description of the app to the Project section.

Reviewer: ```suggestion
For a more detailed description of the app go to the Project section.[link's missing here]
```

Author: 🍤

## Installation

1. **Prerequisites:** The project needs [Node](https://nodejs.org) ^10.13.0 and [Yarn](https://yarnpkg.com) ^1.12.3 to as a dependency manager.
2. Download or clone the project to a local folder
3. While in the project folder run `yarn install`

> 4. When all dependencies are installed, run `yarn start` to start the dev server

Reviewer: ```suggestion
4. After all dependencies are installed, run `yarn start` to start the dev server
```

Author: 🍤

### The Project

### Features

-   Show a list of dogs retrieved from [DogAPI](https://dog.ceo/dog-api/)
-   Filter the list by dog breeds through a text input
-   Show an image of each breed retrieved from [DogAPI](https://dog.ceo/dog-api/)
-   Each dog breed name appears on a tooltip when you hover over the image
-   Each dog has its own page with details retrieved from [MediaWiki](https://www.mediawiki.org/wiki/API:Main_page)
-   Submit a dog picture through a form

> -   Each dog card is only rendered when the component is in the viewport, by using a lazy load component

Reviewer: ```suggestion
-   Each dog card is only rendered when the component is in the viewport, using a lazy load component
```

Author: Done.

-   Each dog image has a preload of the image and displays a placeholder if there were loading errors

### Architecture

-   The project uses React for UI rendering
-   Redux is used as a state manager for sharing state between non immediate parent/child components
-   A `<Router />` component was created from scratch based on the [react-router](https://github.com/ReactTraining/react-router) implementation

> -   The router components use **react context** as a state manager instead of redux as the router needed the easiest way to pass data between components

Reviewer: I don't think the part I removed below is really necessary, you could just say in the previous line that Redux was implemented for the sake of learning it. I'd say that keeping state management in your router free of 3rd party libraries (aka using just React instead of Redux) is a better reason to list.

Thoughts?

```suggestion
-   The router components use **react context** for state management
```

Reviewer: A better option is not even mentioning this detail here, maybe only in the Router and State Management sections.

Author: I think this is mentioned as you said in the Router Components section, if not I'm going to add it there and delete the explanation in this point.

#### State Management

> As the app kept getting more features passing data between components became more complicated, as a way to simplify the state passed between components, Redux was implemented. This can be seen in the `<SubmitDog />` and `<AllDogs />` components which share the same data.

Reviewer: ```suggestion
As the app kept getting more features, passing data between components became more complicated. Redux was implemented as a way to simplify the state passed between components. This can be seen in the `<SubmitDog />` and `<AllDogs />` components which share the same data.
```

Author: Done.

> As Redux's actions don't wait asynchronous functions, the [redux-thunk](https://github.com/reduxjs/redux-thunk) middleware was used for resolving this behavior.

Reviewer: ```suggestion
As Redux' actions are not asynchronous out of the box, the [redux-thunk](https://github.com/reduxjs/redux-thunk) middleware was used for resolving this behavior.
```

[Words that finish in s, x or z don't use 's to show possession](http://www.plainenglish.co.uk/using-apostrophes.html). Only the apostrophe is used.

Author: Done.

In this app, Redux handles the fetching and sending of data from the APIs as well as managing this information in the store. It also stores and removes messages from the message list that `<MessageHandler />` uses.

### Key parts

#### `<App />`

The root component of the App.
Connects with the Redux store and actions to be passed as props.

> On mount, the component fetches the dog's list from local storage if present or from the API to be set on store.

Reviewer: Is this data fetched by the component or Redux?

Author: Call an action that retrieves the list, going to change the wording.

#### Router

The `<Router />` component was coded from scratch based on the [react-router](https://github.com/ReactTraining/react-router) library. It has four main components: `<Router />`, `<Route />`, `<Redirect />` and `<Otherwise/ >`. There's also a `<Link />` component to assist with navigation.
Uses **React Context** to handle the state and pass it to components that may need it.

[Read more about the router](./components.md#router)

> #### ShowDogs Components

Reviewer: Didn't this change to DogsList of something like that?

Author: The component is DogList but the folder with all the components regarding the showing of dogs is called ShowDogs, thats why I don't put it as <ShowDogs />. It would be better to just title the section as DogList Components?

The main content of the app. Several components that fetch a list of dog breeds and manages image display of them.

> Shows as a card list, with the image of each dog showing at first, but when hovering on each image, a tooltip created with **React Portals** follows the cursor with the breed name until it leaves the image.

Reviewer: ```suggestion
Shows as a card list, with the image of each dog. On hover, displays a tooltip created with **React Portals** that follows the cursor with the breed name until it leaves the image.
```

Author: Done.

A lazy loader was implemented to load the image only when the component is inside the viewport, meanwhile or if there are any errors, a placeholder will be shown.

> The dog list can be filtered by breed using a controlled component text input.

Reviewer: ```suggestion
The dog list can be filtered by breed using a controlled text input.
```

> Each dog has its own page to display some information fetched from the Wikipedia Media API, as this part was just an excuse to populate a page with different text, some information that the API returns may not be really related to the dog breed.

Reviewer: ```suggestion
Each dog has its own page to display some information fetched from the Wikipedia Media API. Information from the API may not be really related to the dog breed, as implementing this was beyond the scope of this project at the moment.
```

Author: Done.

> A route component was created to only show an image if the breed of the dog page is a Pug, this was used to test the route component as not a direct child of the router component.

Reviewer: ```suggestion
When on the Pug's detail page (/whatever/pug), a <Route> component is used to display a specific section.
```

Author: Done.

[Read more about the showdogs components](./components.md#showdogs)

#### `<SubmitDog />`

**At the moment this component is still under work and only showcases error handling**

Allows the user to submit a dog image with the name and breed of the dog. The `<select>` element lists all the dogs' breeds received from the API and passed down by `<App />`.
An image preview is shown next to the form after the user selects an image from their computer.
Form submission calls an action that sends the form data to the corresponding endpoint, which currently don't exists and will be developed in the second part of this app. For now, it just showcases error handling.

[Read more about SubmitDog](./components.md#submitDog)

#### Error Handling

To display alerts, warning, errors and any necessary type of message, the app uses `<MessageHandler />` and `<Snackbar />` components.

> The `<MessageHandler>` receives a list of messages which renders as `<Snackbars />` until they are either manually or automatically closed. If there are several message at the same time they start to stack up from bottom to top. When the `<Snackbar />` component is closed a callback function can be called to perform any other action, for example, removing the message from state.

Reviewer: ```suggestion
The `<MessageHandler>` receives a list of messages which renders as `<Snackbars />` until they are either manually or automatically closed. If there are several messages at the same time they start to stack up from bottom to top. When the `<Snackbar />` component is closed, a callback function can be called to perform any other action, for example, removing the message from state.
```

Author: Done.

> The `<Snackbar />` component shows either a string message or any given children. The component alone is able to autoclose after some time, defaults to 3000ms but can also be passed as a prop for diferent times. There are other options like not autoclosing, hiding the close button, and the type of message to be shown (alert, warning, error).

Reviewer: ```suggestion
The `<Snackbar />` component shows either a string message or any given children. The component alone is able to autoclose after some time, defaults to 3000ms but can also be passed as a prop for different times. There are other options like not autoclosing, hiding the close button, and the style to be used for the message box (alert, warning, error).
```

Author: Done.

[Read more about the Error Handling components](./components.md#error-handling)

### Disclaimer

Some of the app features may be incomplete or can be enhanced. The focus of this project was not to have a polished off application, but to learn, comprehend and implement modern tools and processes.

## Testing

> Testing was implemented on late stage of the UI development, in parallel to Redux. All of Redux's actions and reducers were implemented using **Test Driven Development**. The `<SubmitDog />` component was tested after learning how to unit test on Redux. After this, the `<MessageHandler />` and `<Snackbar />` components were created using TDD.

Reviewer: ```suggestion
Testing was implemented on a late stage of the UI development, in parallel to Redux. All of Redux' actions and reducers were implemented using **Test Driven Development**. The `<SubmitDog />` component was tested after learning how to unit test on Redux. After this, the `<MessageHandler />` and `<Snackbar />` components were created using TDD.
```

Author: Done.

As testing was implemented late in the project, not all components have tests.

To test the app in the project folder run `yarn test`.

#### Tools

-   **[Jest](https://jestjs.io/)** was used as the main testing platform for the app.
-   **[Enzyme](https://airbnb.io/enzyme/docs/guides/jest.html)** library created by AirBnB to test components
-   **Mock libraries** used:
    -   [mock-local-storage](https://www.npmjs.com/package/jest-localstorage-mock) mock for testing localstorage
    -   [mock-fetch](https://www.npmjs.com/package/fetch-mock) mock for testing dog list and dog image fetching
    -   [mock-store](https://github.com/dmitry-zaets/redux-mock-store) redux store mock for async actions and middleware

## Components

[See a detailed list of the components used in the project.](./components.md)

## Styles

-   Styles were added using a vanilla stylesheet that loads inside `<App />`.
-   [Bootstrap 4](https://getbootstrap.com/) is loaded from official CDN on index.html

## Conclusion

> The project is teaching me how to use modern tools and processes such as **React**, **Redux**, **Test Driven Development** using **Jest** as a testing platform and in the next step of the project I will learn **NodeJS**.

Reviewer: ```suggestion
The project is taught me how to use modern tools and processes such as **React**, **Redux**, **Test Driven Development** using **Jest** as a testing platform and in the next step of the project I will learn **NodeJS**.
```

Author: Done.

The project was started without a boilerplate, I learned how to start up a project from scratch using **Yarn**, **Webpack** and **Babel**. This project was focused on React but this knowledge can be used for any kind of project.

> Git is used, which I have already known and use before but this is the first time I use **git-flow** in a project.

Reviewer: ```suggestion
Git is used, which I already knew and used before, but this is the first time I use **git-flow** workflow in a project.
```
Would you use it again? why? Why was it helpful or not?

Author: Done.

After configuring the project, I learned how to plan the architecture of an app, defining specs and features and how they should interact with the application and between them.

> As the coding started, there was so much to learn in react, starting with JSX, components and their lifecycle, props, state, the render method, render props, high order components, refs, portals, context... to name some. My mind set also changed as I have to start thinking in React. Immutability, components just serving one purpose, when and when not to create a state, how components should communicate between each others.

Reviewer: ```suggestion
As the coding started, there was so much to learn in react, starting with JSX, components and their lifecycle, props, state, the render method, render props, high order components, refs, portals, context... to name some. My mindset also changed as I have to start thinking in React. Immutability, components just serving one purpose, when and when not to create a state, how components should communicate between each other.
```

Author: Done.

> When the app state management necessities grew, Redux was implemented with Test Driven Development process using Jest as the testing platform. This was new for me, I learned a lot from this as to how to set up a testing environment, which helpers I would need to use and how to use them, and what to test in the code. Then some components were created using TDD with the help of the **Enzyme** library.

Reviewer: ```suggestion
When the app state management necessities grew, Redux was implemented with Test Driven Development process using Jest as the testing platform. This was new for me, I learned a lot from this: how to set up a testing environment, which helpers I would need to use and how to use them, and what to test in the code. Then some components were created using TDD with the help of the **Enzyme** library.
```

Author: Done.

> Redux actions were used to fetch information from an API, so I learned to use **Thunk** as a middleware in this cases. I learned actions and reducers and how they connect to the app passing the store and actions as props to the components that needed them.

Reviewer: ```suggestion
Redux actions were used to fetch information from an API, so I learned to use **Thunk** as a middleware in these cases. I learned what actions and reducers are, and how they connect to the app, passing as props to the components that needed them.
```
These is used as the plural of this.

Author: Done.

> #Learning React with Dogs

Reviewer: There should be an empty space between `#` in titles and the first word, otherwise they won't be rendered properly.
![screenshot_20190103_094257](https://user-images.githubusercontent.com/1082158/50646594-1a95f380-0f3c-11e9-898e-1741e4d20052.png)

Reviewer: Also, remember to put all component names between `<Name />` to specify when you're talking about a component.

Author: Done!

=========================

### What is this?

> As the name implies is a personal project as a way to learning while coding.

Reviewer: ```suggestion
As the name implies, this is a personal project, meant for learning while coding.
```

Author: Done!

> The main focus of learning was React, Redux, Test Driven Development with Jest and as a final step NodeJs.

Reviewer: ```suggestion
The main focus was to learn React, Redux, Test Driven Development with Jest and eventually NodeJS (TBD).
```
Node is not done yet, so make a note about it, there is no node on this project yet.

Author: Done!

> ### What the app does?

Reviewer: ```suggestion
### What is the app's purpose?
```

> The app manages data from [DogAPI](https://dog.ceo/dog-api/) and shows it in a simple way in the app.  

Reviewer: ```suggestion
The app manages data from [DogAPI](https://dog.ceo/dog-api/) and shows it in a simple way.
```

I feel like this is too short of a description, also, I don't like the word "manage" it doesn't manage data at all, just fetches and shows it.

Author: Done!

> For more detailed description of the app specs go to the App and Components description section.

Reviewer: ```suggestion
For a more detailed description of the app go to the Project section.
```
Also, I feel like this section should be the top part of the Project section, not a standalone section.

Author: Done!

Installation
------------

1. **Prerequisites:** The project needs [Node](https://nodejs.org) ^10.13.0 with [NPM](https://nodejs.org) ^6.4.1 (included with Node) or [Yarn](https://yarnpkg.com) ^1.12.3 to run the project.
2. Download or clone the project to a local folder

> 3. While in the proyect folder run `yarn install` or `npm install` depending what you use

Reviewer: ```suggestion
3. While in the project root, run `yarn install`
```
YOU and more importantly, this project are using yarn, it uses a yarn.lock file, do not suggest the use of npm at all.

Author: Done!

> 4. When all dependencies are installed, run `yarn start` or `npm start` to start the dev server

Reviewer: Same here, remove all npm references.

Author: Done!

###The Project
-----------

###Features
 
 - Show a list of dogs retrieved from [DogAPI](https://dog.ceo/dog-api/)

>  - Filter the list by the breed of the dog through a user input

Reviewer: ```suggestion
 - Filter the list by dog breeds through a text input
```

Author: Done!

>  - Show an image of each dog breed retrieved from [DogAPI](https://dog.ceo/dog-api/) *lazydog

Reviewer: ```suggestion
 - Show an image of each breed retrieved from [DogAPI](https://dog.ceo/dog-api/)
```

Author: Done!

>  - Each dog breed name appears on a tootltip when you hover over the dog

Reviewer: ```suggestion
 - Each dog breed name appears on a tootltip when you hover over the image
```

Author: Done!

>  - Each dog has his own page with details about the dog retrieved from [MediaWiki](https://www.mediawiki.org/wiki/API:Main_page)

Reviewer: ```suggestion
 - Each dog has its own page with details retrieved from [MediaWiki](https://www.mediawiki.org/wiki/API:Main_page)
```
If you're not referring to a specific dog, use "it" instead of a gender pronoun.

Author: Done, thanks!

 - Submit a dog picture through a form

>  - Each dog card is only rendered when the component is in the viewport sight (lazy load)

Reviewer: ```suggestion
 - Each dog card is only rendered when the component is in the viewport, by using a lazy load component
```

Author: Done!

>  - Each image has a preload of the image and report if there was errors on the loading

Reviewer: I don't really understand what you're trying to say here with the preload, but I have a suggestion for the last part.

```suggestion
 - Each image has a preload of the image and displays a message if there were loading errors
```

Author: 🍤

###Architecture

> - The project uses react as a framework

Reviewer: React is a UI library, you can leave it as:

```suggestion
- The project uses React for UI rendering
```

Author: 🍤

> - Redux is used as a state manager although the project was started without it

Reviewer: People do not care if it was started without it, most projects are, they will want to know WHY it was used.

Author: 🍤

> - A router was created from scratch based on the [react-router](https://github.com/ReactTraining/react-router) library

Reviewer: ```suggestion
- A `<Router>` component was created from scratch based on the [react-router](https://github.com/ReactTraining/react-router) implementation
```

Author: 🍤

> - The router components use **react context** as a state manager instead of redux for learning purposes

Reviewer: Context was not used here just because of learning purposes, Context is the easiest way to pass data between components. It was used because it was necessary.

Author: 🍤

###Key parts
------------

> ####App

Reviewer: Write component names as JSX components to make explicit what they are.
```suggestion
#### <App />
```

Same for all the others below.

Author: 🍤

> The main component of the App, is the parent of all teh components in the app.

Reviewer: ```suggestion
The root component of the App.
```
It is understood that all other components will be children of this one.

Author: 🍤

> App component connects with the Redux store and actions to be passed as props.

Reviewer: ```suggestion
Connects with the Redux store and actions to be passed as props.
```
This is the App section, no need to repeat it.

Author: 🍤

> On mount the component fetches the dog list from local storage or from the api depending the case to be setted on store.

Reviewer: Redaction can be improved.

```suggestion
On mount, the component fetches the dog's list from local storage if present, or from the API, to be set on store.
```

Author: 🍤

> This component renders the **Router** component with the main routes, redirects and otherwise. The component also renders the **MessageHandler** component which shows all the messages through the app.

Reviewer: None of this is really important here. Both have their own section, and it is implied that **anything** will be rendered inside `<App />`

Author: 🍤

####Router

> As mentioned before a **Router** component was coded from scratch based on the [react-router](https://github.com/ReactTraining/react-router) library for managing routing. The router is composed by 4 components: **Router**, **Route**, **Redirect** and **Otherwise**. There's also a link component to use as `<a>` element that implements link routing.

Reviewer: ```suggestion
The **<Router />** component was coded from scratch based on the [react-router](https://github.com/ReactTraining/react-router) library. It has four main components: **Router**, **Route**, **Redirect** and **Otherwise**. There's also a `<Link />` component to assist with navigation.
```

Also, remember to put all components inside `</>`

Author: 🍤

> The router is the ony part of the app that uses **React Context** to handle the state passed to the link component.

Reviewer: ```suggestion
Uses **React Context** to handle state and pass it to components that may need it.
```

Author: 🍤

[Read more about the router](###Router)

####ShowDogs

> The main content of the app. Several components that fetch a list of breed of dogs and images of each breed.

Reviewer: You're mixing up two things here, one is the list of components that display an image, the other is fetching the list data. It's a not so subtle difference, but it is important since this component doesn't really do both.

> The list shows as a card list, with ony the image of each dog showing at first instance but when hovering the mouse on each image a tooltip created with **React Portals** follows the mouse with the dog's breed name until it lefts the image.

Reviewer: Wording can be greatly improved in general, please take note of the suggestions.

```suggestion
Shows card list, with the image of each dog showing at first, but when hovering on each image, a tooltip created with **React Portals** follows the cursor with the breed name until it leaves the image.
```

> A lazy loader was implemented to only render the dogs on list when the image is on the viewport sight and it starts fetching the dog's image which starts preloading with a placeholder image that changes if the image loads without errors.

Reviewer: You're repeating yourself by saying "viewport sight".

```suggestion
A lazy loader was implemented to load the image only when the component is inside the viewport, meanwhile or if there are any errors, a placeholder will be shown.
```

Author: 🍤

Reviewer: :fried_shrimp:

> The dog list can be filtered by the dog's breed using a **controlled component** with an input text.

Reviewer: ```suggestion
The dog list can be filtered by breed using controlled text input.
```

Author: 🍤

> Each dog has his own page that shows a little information fetched from Wikipedia Media Api, as this part was just an excuse to populate a page with different text, some information that the API returns may be different of what is expected with the dog breed. A route component was created to only show an image if the breed of the dog page is a Pug, this was used to test the route component as not a direct child of the router component.

Reviewer: ```suggestion
Each dog has its own page to display some information fetched from the Wikipedia Media API, as this part was just an excuse to populate a page with different text, some information that the API returns may not really be related to the dog breed.

To showcase conditional rendering and passing context based on routing, a <Route> component is used to show an image for the **Pug** page. (PUT THE RIGHT PATHNAME HERE)
```

Author: 🍤

[Read more about the dog components](###Showdogs)

> ####SubmitDog

Reviewer: Specify from the start that form submission doesn't work and for now it just showcases error handling.

Author: 🍤

> This component allows the user to submit a dog image with the name and breed of the dog. The `<select>` element listing all the dogs breeds is populated with the dog list fetched from app component.

Reviewer: ```suggestion
Allows the user to submit a dog image with the name and breed of the dog. The `<select>` element lists all the dogs' breeds received from the API and passed down by `<App/>`.
```

Author: 🍤

> When an image is seleted from the file field a preview of it is showed besides the form.

Reviewer: ```suggestion
An image preview is shown next to the form after the user selects an image from their computer.
```

Author: 🍤

> When submitting the form an action is called that should send the form data to the corresponding endpoint, which currently don't exists and it's gonna be the second part of this app.

Reviewer: ```suggestion
Form submission calls an action that sends the form data to the corresponding endpoint, which currently don't exists and will developed in the second part of this app. For now, it just showcases error handling.
```

Author: 🍤

[Read more about SubmitDog](###SubmitDog)

####MessageHandler and Snackbar

> The app have the need of a message handler component to show all the alerts, warning, errors and any necessary type of message. The MessageHandler component recieves a list of messages which renders as Snackbars and shows them into the app until the snackbar is closed. If there's several message at the same time they start to stack up from bottom to top. When the snackbar component is closed a callback removes the message from the list.

Reviewer: This section should be called Error Handling.

```suggestion
To display alerts, warning, errors and any necessary type of message, the app uses the `<MessageHandler />` and `<Snackbar />` components.

The `<MessageHandler> receives a list of messages which renders as` <Snackbars />` until they are either manually or automaticalle closed. If there are several message at the same time they start to stack up from bottom to top.
When the `<Snackbar />` component is closed a callback function can be called to perform any other action, for example, removing the message from state.
```

Please note that the callback **does not remove the message from the list**, it just enables the user to implement it.

Author: 🍤

> The Snackbar component shows the message as a passed string prop or as the component children. The component alone is able to autoclose after some time, defaults to 3000ms but can also be passed as a prop for diferent times. There's also other options as not autoclosing (passing 0ms as a time closes immediately), showing or not the close button and the type of message to be shown (alert, warning, error).

Reviewer: ```suggestion
The Snackbar component shows either a string message or any given children components. The component alone is able to autoclose after some time, defaults to 3000ms but can also be passed as a prop for diferent times. There's also other options like not autoclosing, or hiding the close button and the type of message to be shown (alert, warning, error).
```

Autoclosing at 0 seems like a bug by design, I would not mention it tbh.

Author: 0

[Read more about MessageHandler and Snackbar components](###MessageHandler)

> ####Redux

Reviewer: This should be up top, on the Architecture section, and with a State management title instead of just Redux. Remember, you're not learning Redux, you're learning State Management through Redux.

Author: 🍤

> As the app kept getting more features a way to manage the state between not parent/children components was an increasing necessity as in the submitDog component needed the list of dog breeds. Redux was implemented with the help of the [react-redux](https://react-redux.js.org/) bindings.

Reviewer: ```suggestion
As the app kept getting more features, passing data between components became more complicated, as a a way to simplify the state passed between components, Redux was implemented. This can be seen in the <SubmitDog /> and <ShowDogs /> components, which share the same state data.
```

Author: 🍤

> Actions and reducers were created for functionality as fetching and setting the dog list and images, submitting a dog from the form, and adding or removing messages from the message list in store.

Reviewer: You can assume someone reading this may know Redux, hence, they know actions and reducers were created for functionality, this can be rephrased to mention the features that were delegated to Redux.

For example: "Redux handles fetching data from the API, as well as sending data, storing error messages and bla bla bla"

Author: 🍤

> As actions don't dispatch asynchronous functions, [redux-thunk](https://github.com/reduxjs/redux-thunk), a middleware for resolving this was used in this cases.

Reviewer: ```suggestion
As Redux' actions don't wait asynchronous functions, the [redux-thunk](https://github.com/reduxjs/redux-thunk) middleware was used to enable this behavior.
```

Author: 🍤

> ### Disclaimer

Reviewer: I'd put this up top, probably.

Author: 🍤

> As the main objective of the project was to learn to use certain tools in modern development in a real project, many of the app features may be incomplete or can be greatly enhaced. So the focus of this project more than having a polished off app is to comprehend and implement modern code with moderns tools. 

Reviewer: ```suggestion
Some of the app features may be incomplete or can be enhanced. The focus of this project was not to have a polished off application, but to learn, comprehend and implement modern tools and processes.
```
I don't think they are thaaaat incomplete, but they can definitely be enhanced.

Author: 🍤

##Testing
-------

> Testing was implemented on late state of the app when redux was also beign implemented. All of redux actions and reducers were implemented using **Test Driven Development**, writing the tests first and then the code or functions that this were testing.

Reviewer: ```suggestion
Testing was implemented on late stage of the UI development, in parallel to Redux. All of Redux' actions and reducers were implemented using **Test Driven Development**.
```

Author: 🍤

> When redux was implemented, the already created SubmitDog component was also tested as an example on how to test components. MessageHandler and Snackbar components were created using TDD after this.

Reviewer: ```suggestion
The `<SubmitDog />` component was tested after learning how to unit test on Redux. After this, the `<MessageHandler />` and `<Snackbar />` components were created using TDD.
```

Author: 🍤

> As testing was implemented late in teh project and only as a mean to learn, most components remain untested and the code and component for which testing was written may need more testing cases.

Reviewer: ```suggestion
As testing was implemented late in the project, not all components have tests.
```

Author: 🍤

> To test the app in the project folder run `yarn test` or `npm test`.

Reviewer: ```suggestion
To test the app in the project folder run `yarn test`.
```

Author: 🍤

####Tools
- **[Jest](https://jestjs.io/)** was used as the main testing platform for the app.
- **[Enzyme](https://airbnb.io/enzyme/docs/guides/jest.html)** library created by AirBnB to test components

> - Several mocks were used when testing:

Reviewer: ```suggestion
- Mock libraries used:
```

Author: 🍤

	-[mock-local-storage](https://www.npmjs.com/package/jest-localstorage-mock) -mock for testing localstorage  
	-[mock-fetch](https://www.npmjs.com/package/fetch-mock) - mock for testing dog list and dog image fetching
	-[mock-store](https://github.com/dmitry-zaets/redux-mock-store) - redux store mock for async actions and middleware

##Components
--------------

###App
- Main component of the app.
- Connects with redux store

> - Is passed by props store state and redux actions

Reviewer: ```suggestion
- Gets Redux' state and actions as props
```

Author: 🍤

> - On Mount retrieve by localStorage or by fetching the list of dogs from the api

Reviewer: ```suggestion
- Retrieves the dogs list either from localStorage or an API request
```

Author: 🍤

> - Render the main markdown of the app

Reviewer: ```suggestion
- Renders the all the app's content
```

Author: 🍤

- Renders Router components
- Renders MessageHandler component

###Showdogs

> ####AllDogs

Reviewer: This is a minor suggestion, I would rename this component to DogsList or something like it, but feel free to leave it as it is.

Author: 🍤

> - Home content of the app

Reviewer: ```suggestion
- Main section of the app
```

Author: 🍤

> - Shows a list of dog through image cards

Reviewer: ```suggestion
- Display a list of dogs as cards with an image
```

Author: 🍤

> - Filter the dog list in accordance to the filterValue store state

Reviewer: ```suggestion
- Filter the dog list in accordance to the store's filterValue
```

Author: 🍤

####DogFilter
- Form with an input and a reset button
- The reset button only shows when there's text in the input field

> - The form is a directed controlled component so the input recives his value from the store props

Reviewer: ```suggestion
- The form is a controlled component, the input receives its value from store
```

Author: 🍤

- When the user changes the input value an action prop is called to update the value in the store

####LazyDog

> - Checks if the component is in sight of the viewport

Reviewer: ```suggestion
- Checks if the component is inside the viewport
```

Author: 🍤

> - If the card is in sight render the children, if not just render an empty div

Reviewer: Why an empty div? does it have some styling purpose?

Author: 🍤

####DogCard

> - Shows the dog picture as a card

Reviewer: ```suggestion
- Shows the dog picture inside a fixed dimensions box
```

Author: 🍤

> - Have a link to the dog page

Reviewer: ```suggestion
- Links to the dog page
```

Author: 🍤

> - When the mouse is over the card it shows a tooltip created through a react portal that follows the mouse showing the dog's breed name 

Reviewer: ```suggestion
- When the cursor is over the card it shows a tooltip created through a portal. It follows the cursor showing the breed name 
```

Author: 🍤

> - When the mouse leaves the card the tooltip is no longer visible

Reviewer: ```suggestion
- When the cursor leaves the card the tooltip is no longer visible
```

Author: 🍤

####DogPic
- Renders an image element with the picture of the dog

> - Wait for the image to load on a virtual element

Reviewer: What is a virtual element? :thinking:

Author: Was supposed to be a 'draft word idea', mean to think as an element that is created through code but not appended into the body... that is just an element, sorry! 🍤

> - If the image doesn't loads or have errors laoding display a placeholder 

Reviewer: ```suggestion
- Displays a placeholder while the image loads or if there are any errors
```

Author: 🍤

####Dog

> - Shows the dog information through a fetch to wikipedia media api

Reviewer: ```suggestion
- Shows the dog information from the Wikimedia API
```

Author: 🍤

- If the dog breed is a Pug shows an image thorugh route rendering

####DogTip

> - Shows or hides a tooltip that follows the mouse

Reviewer: ```suggestion
- Shows or hides a tooltip that follows the cursor
```

Author: 🍤

> - The tootltip is created by React Portals

Reviewer: ```suggestion
- The tootltip is created with React Portals
```

Author: 🍤

###SubmitDog
- Renders a form to submit a dog image

> - When selected the image to be submitted it shows as a preview

Reviewer: ```suggestion
- Previews the selected image
```

Author: 🍤

> - When the form is submitted sends the data to an action function

Reviewer: ```suggestion
- Calls a Redux action on form submission
```

Author: 🍤

###Commons

####Header
- Renders the header markup

> - Renders the nav menu of the app with router links

Reviewer: ```suggestion
- Renders the navigation menu
```

Author: 🍤

> - Renders the dogFilter component when the route path is the given one

Reviewer: This is confusing, could you try to explain what it means, please?

Author: The <DogFilter /> component only renders when the route is '/dogs' as <DogList /> renders on this same route, but the wording I use is confusing 🤔

####Footer

> - Render the footer, currently an empty `<div>` element

Reviewer: Why render an empty div or the Footer component at all?

Author: It was an error, the footer component should render its children, if there's none then renders nothing

####About

> - Renders an about page with several links

Reviewer: **several**
![inigo](https://user-images.githubusercontent.com/1082158/50654283-f1805d80-0f51-11e9-895f-0737bf642210.jpg)

Author: 🤷‍♂️

####MousePosition

> - Render prop component

Reviewer: ```suggestion
- Receives a render prop to render its children
```

Author: 🍤

> - Pass to the component that renders the mouse position as props

Reviewer: ```suggestion
```

Author: 🍤

###MessageHandler

####Snackbar
- Renders an snackbar component
- The component shows a message passed through props or the component's children
- There are several styles (colors) of snackbars depending on the type
- Auto closes at 3000ms, time can be changed or auto close can be removed 
- There's a hiddeable close button to close the component

> ####MessageHandler

Reviewer: This should be before Snackbar

Author: 🍤

- Renders a div in which messages appear

> - Receives an array of messages and renders each as a snackbar

Reviewer: ```suggestion
- Receives an array of messages as `<Snackbar />` components
```

Author: 🍤

> - Add a callback that removes the message from the list when the snackbar is closed

Reviewer: ```suggestion
- Can take a callback to perform any action after the Snackbar closes
```

Author: 🍤

> ###Router

Reviewer: Router should be much higher, as it is very complex and integral part of the app

Author: 🍤

- Delegates what to show depending on the current location path
- Gets the current location path and divides it into path parts 

> - Have 3 types of routing:

Reviewer: ```suggestion
- Has three routing components:
```

Author: 🍤

	- Router: compares current path with route path and render its children if there's a match

> 	- Redirect: compares current path with redirect path and if it's a match redirect current location to new redirect route path

Reviewer: ```suggestion
	- Redirect: compares the current path with the redirect prop and changes the location to it if there is a match
```

Author: 🍤

> 	- Otherwise: if no route or redirect path is found go to the otherwise route path

Reviewer: ```suggestion
	- Otherwise: redirects to the given path if the current location doesn't match any registered path
```

Author: 🍤

> - Routes, Redirects and Otherwise gets registered in a list of routes

Reviewer: ```suggestion
- The values of Routes, Redirects and Otherwise are registered in the Router's internals
```

Author: 🍤

> - Allows params in the uri to render pages (dog/:dogbreed - dog/shiba)

Reviewer: ```suggestion
- Reads params in the URI to render sub paths (dog/:dogbreed - dog/shiba)
```

Author: 🍤

- Creates a listener on history.pushState to be able to manage routing without having to reload the page

> - This component and the ones related to the router use **React Context** as a state manager for the location, path and parts 

Reviewer: ```suggestion
- Uses React Context to manage and share state
```

Author: 🍤

####Route

> - Register the route to the router

Reviewer: ```suggestion
- Register a path on the router
```

Author: 🍤

- Compares current path or location with the route path and renders its children if there's a match

####Redirect

> - Register the redirect to the router

Reviewer: ```suggestion
- Register a redirect path on the router
```

Author: 🍤

> - Compares current path with redirect route path and if there's a match history push to the redirection

Reviewer: ```suggestion
- Compares the current path with the redirect path and pushes the redirect path to history state when there is a match
```

Author: 🍤

####Otherwise

> - Register otherwise route to the router

Reviewer: ```suggestion
- Register a fallback path on the router
```

Author: 🍤

####Link

> - Renders a link `<a></a>` element

Reviewer: ```suggestion
- Renders a link `<a></a>` element enhanced with the processed routes
```

Author: 🍤

- Gets current location from react context
- Checks if the href attribute of the link is internal or complete URL
- If the href is internal calls for a history.pushState

> - If the href is a complete URL (mostly external links) sets window.location or calls window.open() depending on the link's href attribute

Reviewer: How the link opens should depend on the "target" prop not on the href attribute. It should emulate what a normal anchor element does.

Author: This was a typo mistake, thanks! href -> target

##Styles

> - Styles were added using a vanilla stylesheet that loads on app component.

Reviewer: ```suggestion
- Styles were added using a vanilla stylesheet that loads inside `<App />`
```

Author: 🍤

> - Also [Bootstrap 4](https://getbootstrap.com/) was used in this project which loads from the offical CDN on index.html file

Reviewer: ```suggestion
- [Bootstrap 4](https://getbootstrap.com/) is loaded from offical CDN on index.html
```

Author: :frie

Author: d_shrimp:

> ##Conclusion

Reviewer: This is way below, the components section takes way much space. I'd suggest leaving the Components section at the bottom or even on a different file.

Author: Was just thinking about that, create a Docs folder with a copy of the readme file and passing the Component section o a new file inside Docs

> As the project was made only for learning purposes, it made me code with modern tools on a project and not only a tutorial or sandbox. React with Redux and using Jest for testing is combo that I'm gonna be interesed in using on a real project.

Reviewer: This conclusion is very lacking, I want to know what you learned and how, yeah, sure you can be interested in doing things in the future, but what are those things? What specific things you learned here that will help you in the future? All these concepts have names and I expect to see them here. Try to reread this conclusion as someone looking to hire a dev, does it really tell you anything?

Author: Re-read it. Really lacks. Gonna write it again. :lacked-shrimp:

##Pull Requests and Issues

> As mentioned before, the project was made as means for learning so no pull request are admitted, any issue can be submitted through the [github issue page for the project](https://github.com/luizmendez/learning-react-with-dogs/issues).

Reviewer: ```suggestion
As this project was made for learning purposes, no pull request are taken, but issue reporting is greatly appreciated, and  can be submitted through the [project's issue page](https://github.com/luizmendez/learning-react-with-dogs/issues).
```

Author: 🍤

##Resources
- [React](https://reactjs.org)
- [Redux](https://redux-form.com/)
- [React-Redux](https://react-redux.js.org/)
- [Thunk](https://github.com/reduxjs/redux-thunk)
- [Jest](https://jestjs.io/)
- [Enzyme](https://airbnb.io/enzyme/docs/guides/jest.html)
- [DogAPI](https://dog.ceo/dog-api/)
- [Wikipedia Mediawiki API](https://www.mediawiki.org/wiki/API:Main_page)
- [Bootstrap](https://getbootstrap.com/)
- [NodeJS](https://nodejs.org/)
- [Webpack](https://webpack.js.org/)
- [Yarn](https://yarnpkg.com/en/)

> Special thanks to **[edrpls](https://github.com/edrpls)** for his mentoring during this project.

Reviewer: awww

Author: 😽
